### PR TITLE
Some changes to simplify the generation function

### PR DESCRIPTION
**Pull request by yjernite**
This PR proposes to simplify generation in `modeling_utils.py` in the following ways:

1. Removing some redundant code in `_generate_no_beam_search`: finished sequences are padded at generation time, and do not need to be padded again before returning
2. Initializes the cache in its permanent form directly for both `_generate_beam_search` and `_generate_no_beam_search`: this removes the need for a first step test in `modeling_bart.py` and `modeling_t5.py` (and presumably future cached seq2seq architectures)
3. Took all of the logit post-processing out of `_generate_beam_search` and `_generate_no_beam_search` and put it in a single `finalize_generation_logscores` function that can be used in both instead of duplicating the code

The following slow test pass in addition to the basic suite:
```
RUN_SLOW=1 pytest tests/test_modeling_bart.py
RUN_SLOW=1 pytest tests/test_modeling_gpt2.py
RUN_SLOW=1 pytest tests/test_modeling_t5.py
```

#### Small breaking changes
1. The previous versionof `_generate_no_beam_search` seemed to be adding padding twice. We removed is as is seemed redundant, but noting it here just in case.
2. This PR moves the addition of the CTRL length penalty from before to after the `log_softmax` in `_generate_beam_search`. This changes scores a little bit but apparently doesn't drastically alter model behavior. Per @patrickvonplaten :
>  Now, the only thing that will slightly change with this function is that the CTRL enforce penalty for beam search decoding. I tried the new order out on a couple of tensors and the changes are minimal. Also, the repetition penalty is very hacky anyways and we already changed the function from its original formula of the CTRL paper.

### CODE REVIEW

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -812,6 +812,62 @@ def enforce_repetition_penalty_(self, lprobs, batch_size, num_beams, prev_output
                 else:
                     lprobs[i, previous_token] /= repetition_penalty
 
+    def finalize_generation_logscores(
```
Comment (patrickvonplaten) [inline src/transformers/modeling_utils.py:815]: I like this!

File: src/transformers/modeling_t5.py [hunk 2]
```diff
@@ -1144,11 +1144,7 @@ def forward(
     def prepare_inputs_for_generation(self, input_ids, past, attention_mask, use_cache, **kwargs):
         assert past is not None, "past has to be defined for encoder_outputs"
 
-        # first step
-        if len(past) < 2:
-            encoder_outputs, decoder_past_key_value_states = past, None
-        else:
-            encoder_outputs, decoder_past_key_value_states = past[0], past[1]
+        encoder_outputs, decoder_past_key_value_states = past
```
Comment (patrickvonplaten) [inline src/transformers/modeling_t5.py:1147]: nice!

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -812,6 +812,62 @@ def enforce_repetition_penalty_(self, lprobs, batch_size, num_beams, prev_output
...
+        temperature,
+        do_sample,
+    ):
+        # repetition penalty (from CTRL paper https://arxiv.org/abs/1909.05858)
+        if repetition_penalty != 1.0:
+            self.enforce_repetition_penalty_(
+                next_token_logits, batch_size, num_beams, input_ids, repetition_penalty,
+            )
+
+        if temperature != 1.0:
```
Comment (patrickvonplaten) [inline src/transformers/modeling_utils.py:837]: I guess this makes only sense for sampling or do you think it could / should also be used for greedy / beam search?

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -1299,18 +1346,7 @@ def _generate_no_beam_search(
                     [attention_mask, attention_mask.new_ones((attention_mask.shape[0], 1))], dim=-1
                 )
 
-        # if there are different sentences lengths in the batch, some batches have to be padded
-        if sent_lengths.min().item() != sent_lengths.max().item():
```
Comment (patrickvonplaten) [inline src/transformers/modeling_utils.py:1303]: I think we need this for specific cases - do you think we can get rid of it? @yjernite

Comment (sshleifer) [reply src/transformers/modeling_utils.py:1303]: we already pad above and the tests seem to pass. The bart slow test has an unevenly lengthed batch, so I think we're OK.

File: src/transformers/modeling_encoder_decoder.py [hunk 5]
```diff
@@ -298,7 +298,7 @@ def prepare_inputs_for_generation(self, input_ids, past, attention_mask, **kwarg
 
         # first step
         if type(past) is tuple:
-            encoder_outputs = past
+            encoder_outputs, _ = past
         else:
```
Comment (sshleifer) [inline src/transformers/modeling_encoder_decoder.py:302]: assert it's a tensor?

File: src/transformers/modeling_utils.py [hunk 6]
```diff
@@ -1231,40 +1287,31 @@ def _generate_no_beam_search(
             outputs = self(**model_inputs)
             next_token_logits = outputs[0][:, -1, :]
 
+            next_token_logscores = self.finalize_generation_logscores(
+                next_token_logits,
+                input_ids,
+                batch_size,
+                1,
```
Comment (sshleifer) [inline src/transformers/modeling_utils.py:1294]: comment num_beams (cause it looks wierd :))

File: src/transformers/modeling_utils.py [hunk 7]
```diff
@@ -1231,40 +1287,31 @@ def _generate_no_beam_search(
...
+                eos_token_id,
+                temperature,
+                do_sample,
+            )
+
             # if model has past, then set the past variable to speed up decoding
             if self._use_cache(outputs, use_cache):
                 past = outputs[1]
 
-            # repetition penalty from CTRL paper (https://arxiv.org/abs/1909.05858)
```
Comment (sshleifer) [inline src/transformers/modeling_utils.py:1238]: yay deletion!

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (sshleifer) [reply src/transformers/modeling_utils.py:815]: I would call this 
`postprocess_generation_logscores` cause we have some notion of finalized hypotheses.
But I love this refactor. The `generate` fn was/is way too long to follow and this helps a significant amount.

**Review by sshleifer — APPROVED** [review]
Love this!

File: src/transformers/modeling_utils.py (hunk 6 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_utils.py:1294]: I think we should work with keyword arguments `next_token_logits=next_token_logits`, ... => it's less error prone.

File: src/transformers/modeling_utils.py (hunk 4 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_utils.py:1303]: @sshleifer Note that this is part of `_generate_no_beam_search`, so it is not tested in Bart. Also, it doesn't matter if the `input_ids` for the encoder have an unevenly lengthed batch since generation only concerns with the decoder.

I looked a bit more into it and thought of a bunch of cases and I agree that I don't think this is necessary. I remember vaguely though that I kept it for some reason to capture a special case, but maybe I was just wrong before. 

So good for me to remove this for now.

Also it's important to note that the tests are randomized (since they use sampling, so it might very well be that the tests are fine in one run, but not anymore in the other one). So in case we have flaky generation tests, this could be a reason. Also linking the PR that added this chunk of code for reference: https://github.com/huggingface/transformers/pull/2885.
It also seems to me that we can remove it

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_utils.py:815]: @yjernite - Actually taking a closer look it this, I'm not so sure if we even should follow suite with a unified sampling function. As it is implemented now, `_generate_no_beam_search` applies two softmax operations (log-softmax and softmax) instead of only softmax (as it was before) so this is a big breaking change for all models doing sampling (GPT2 generation). In beam search, sampling is very rarely used, but in `no_beam_search` it is heavily used and I'm not sure we want a breaking change here. 

The tests don't capture that because we don't have a sampling test for GPT2 only a greedy test (we should prob add a sampling test for GPT2).

File: src/transformers/modeling_utils.py [hunk 8]
```diff
@@ -812,6 +812,62 @@ def enforce_repetition_penalty_(self, lprobs, batch_size, num_beams, prev_output
...
+            self.enforce_repetition_penalty_(
+                next_token_logits, batch_size, num_beams, input_ids, repetition_penalty,
+            )
+
+        if temperature != 1.0:
+            next_token_logits = next_token_logits / temperature
+
+        if self.config.is_encoder_decoder and do_sample is False:
+            # TODO (PVP) still a bit hacky here - there might be a better solution
+            next_token_logits = self.prepare_logits_for_generation(
```
Comment (patrickvonplaten) [inline src/transformers/modeling_utils.py:842]: This is usually only used for beam search (maybe not that important because it is only used for Bart anyways)

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_utils.py:815]: Okey that took me some time to realize :D So because `softmax(log_softmax(X))` is equal to `softmax(X)`, there is not even a breaking change here. But now we calculate the softmax twice instead of only once as before, is the extra cost that insignificant @yjernite ?

**Comment by patrickvonplaten** [comment]
Okey I took a deeper look at the PR! Quite hard to see what's all going on there, so a nice refactoring would be very welcome :-). 

I very much like the small changes that are done here that clearly improve the readability and clean the code.

The big change of unifying the score computation I am not yet 100% convinced it helps a lot. I agree that we should refactor the generate() function, but not sure whether this is going in the right direction. 
+ 
The function reduces duplicated code
- 
The function adds more computational cost to the `.generate()` function, which is already quite expensive and heavily used in GPT2 by an additional softmax function. Thinking about how big the output embedding matrices are, this could be significant no?

I do agree though that we should unify all these functions that are applied to the scores.
IMO, we have to very careful with everything that touches sampling in `_no_generate_beam_search` (greedy decoding is less used here) and everything that touches `argmax` in `_generate_beam_search` (summarization and translation rely on that).

My proposal would be the following: 
Let's unify all functions that are applied after the `F.log_softmax(next_token_logits, dim=-1)` line into one function and that expects the scores (I like @sshleifer's naming - I would just say `scores` instead of `logscores`):

```def postprocess_generation_scores(
        self,
        scores,
        input_ids,
        batch_size,
        num_beams,
        repetition_penalty,
        no_repeat_ngram_size,
        bad_words_ids,
        cur_len,
        min_length,
        max_length,
        eos_token_id,
        temperature,
        do_sample,
    ):
        # repetition penalty (from CTRL paper https://arxiv.org/abs/1909.05858)
        if repetition_penalty != 1.0:
            self.enforce_repetition_penalty_(
                scores, batch_size, num_beams, input_ids, repetition_penalty,
            )

        # Temperature does not influence beam search or greedy search
        if temperature != 1.0 and do_sample is True:
            scores = scores / temperature

        # set eos token prob to zero if min_length is not reached
        if eos_token_id is not None and cur_len < min_length:
            scores[:, eos_token_id] = -float("inf")

        if no_repeat_ngram_size > 0:
            # calculate a list of banned tokens to prevent repetitively generating the same ngrams
            num_batch_hypotheses = batch_size * num_beams
            # from fairseq: https://github.com/pytorch/fairseq/blob/a07cb6f40480928c9e0548b737aadd36ee66ac76/fairseq/sequence_generator.py#L345
            banned_batch_tokens = calc_banned_ngram_tokens(
                input_ids, num_batch_hypotheses, no_repeat_ngram_size, cur_len
            )
            for i, banned_tokens in enumerate(banned_batch_tokens):
                scores[i, banned_tokens] = -float("inf")

        if bad_words_ids is not None:
            # calculate a list of banned tokens according to bad words
            banned_tokens = calc_banned_bad_words_ids(input_ids, bad_words_ids)

            for i, banned_tokens in enumerate(banned_tokens):
                scores[i, banned_tokens] = -float("inf")

        return scores
```

**Comment by patrickvonplaten** [comment]
The temperature function should then for both beam search and no beam search only be applied in the `if do_sample=True` statement (no need to do this for argmax).

Now, the only thing that will slightly change with this function is that the CTRL enforce penalty for beam search decoding. I tried the new order out on a couple of tensors and the changes are minimal. Also the repition penalty comes from the CTRL paper and is not heavily used (We don't use it in beam search as far as I know @sshleifer).

**Comment by sshleifer** [comment]
`repetition_penalty!=1` only used by CTRL and `model_cards/mrm8488/t5-base-finetuned-summarize-news/README.md:` So I think minimal changes to when it is computed are fine.

**Comment by yjernite** [comment]
I changed the `finalize_logits` to @patrickvonplaten 's suggested `postprocess_next_token_scores`

Following @sshleifer 's sage advice, I'm leaving the BART starting hack and the temperature in the main body of the generate function for now, will leave dealing with those for a future PR :)

File: src/transformers/modeling_utils.py [hunk 9]
```diff
@@ -1231,40 +1274,32 @@ def _generate_no_beam_search(
             outputs = self(**model_inputs)
             next_token_logits = outputs[0][:, -1, :]
 
+            scores = self.postprocess_next_token_scores(
+                next_token_logits,
```
Comment (patrickvonplaten) [inline src/transformers/modeling_utils.py:1278]: can we use keyword arguments here since we have that many inputs to make it more robust:
```python
input_ids=input_ids,
no_repeat_ngram_size=no_repeat_ngram_size,
...
```

File: src/transformers/modeling_utils.py [hunk 10]
```diff
@@ -1372,50 +1396,35 @@ def _generate_beam_search(
...
-            if repetition_penalty != 1.0:
-                self.enforce_repetition_penalty_(
-                    next_token_logits, batch_size, num_beams, input_ids, repetition_penalty,
-                )
+            scores = F.log_softmax(next_token_logits, dim=-1)  # (batch_size * num_beams, vocab_size)
 
-            if temperature != 1.0:
-                next_token_logits = next_token_logits / temperature
+            scores = self.postprocess_next_token_scores(
+                scores,
```
Comment (patrickvonplaten) [inline src/transformers/modeling_utils.py:1402]: also keyword arguments here would be nice.

**Comment by patrickvonplaten** [comment]
Great, I'm happy with the PR - I think it already makes generate a bunch more readable. 
Can we note in the PR description, that we have slight breaking changes for beam search sampling when running with the repetition penalty? and it would be nice to make the function call more robust by using keyword arguments the same way it is done with `_no_beam_search_generate()`

**Comment by yjernite** [comment]
> Great, I'm happy with the PR - I think it already makes generate a bunch more readable.
> Can we note in the PR description, that we have slight breaking changes for beam search sampling when running with the repetition penalty? and it would be nice to make the function call more robust by using keyword arguments the same way it is done with `_no_beam_search_generate()`

Done and done, will merge today.

**Comment by sshleifer** [comment]
In future, we also need to run 
```bash
RUN_SLOW=1 tests/test_modeling_marian.py
```
